rosidl_generator_cpp/include/rosidl_generator_cpp/MessageTypeSupport.h

```diff
@@ -17,13 +17,22 @@
 #define __rosidl_generator_cpp__MessageTypeSupport__h__
 
 #include <rosidl_generator_c/message_type_support.h>
+#include <rosidl_generator_c/visibility_control.h>
 
 namespace rosidl_generator_cpp
 {
 
 template<typename T>
+ROSIDL_PUBLIC
 const rosidl_message_type_support_t * get_type_support_handle();
 
 }  // namespace rosidl_generator_cpp
 
+// Tail include the implementation for this function.
+// This include is provided by the implementation and therefore only
+// one version of this header will get included from the path.
+// This header maps the get_type_support_handle function to an implementation
+// specific version of that function.
+#include <rosidl_generator_cpp/get_type_support_handle_impl.hpp>
```

Comment:
I think we can easily keep the advantages you describe:

* single include for user land code
* fail nicely when no implementation is available

but still avoid the disadvantage of the current approach:

* header of this package is not includable
* header of this package references a file which doesn't exist (within the scope of this package and its dependencies)

This can be done by:

* user land includes the impl header
  * if no implementation is on the include path it will pick up the impl header file from this package giving a really good error using `#error`
  * otherwise use the impl header of the implementation package, this had can itself include the header from this package containing the forward declaration